### PR TITLE
[serve] Long polling for routes in http server

### LINKED ISSUE #12302: [serve] Long polling for HTTP/endpoint config

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

Also some cleanup to use constants for the long poll keys.

## Related issue number

Closes https://github.com/ray-project/ray/issues/12302

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/constants.py [hunk 1]
```diff
@@ -37,3 +37,15 @@
...
+LONG_POLL_KEY_REPLICA_HANDLES = "replica_handles"
+
+#: Long poll key for traffic policies.
+LONG_POLL_KEY_TRAFFIC_POLICIES = "traffic_policies"
+
+#: Long poll key for backend configs.
+LONG_POLL_KEY_BACKEND_CONFIGS = "backend_configs"
+
+#: Long poll key for route table.
+LONG_POLL_KEY_ROUTE_TABLE = "route_table"
```
Comment (simon-mo) [inline python/ray/serve/constants.py:51]: Can we use enum for this? Group them in `enum` is definitely more maintainable for future than prefix everything with LONG_POLL_

File: python/ray/serve/tests/test_api.py [hunk 2]
```diff
@@ -29,6 +29,7 @@ def function(flask_request):
     timeout_sleep = 0.5
     while True:
         try:
+            print("1")
```
Comment (simon-mo) [inline python/ray/serve/tests/test_api.py:32]: remove?

**Review by simon-mo — COMMENTED** [review]
There might be a lot of flaky tests because we typically do a single requests.get after client.create_endpoint. Should we change the create_endpoint call to block until the http route is ready? (by checking `/-/routes` in the http server)

File: python/ray/serve/constants.py (hunk 1 above)
Comment (edoakes) [reply python/ray/serve/constants.py:51]: good idea

**Comment by edoakes** [comment]
@simon-mo comments addressed, ptal

**Comment by edoakes** [comment]
@simon-mo I added some code to `create_endpoint` on the client side that will wait for the route to be propagated.

File: python/ray/serve/api.py [hunk 3]
```diff
@@ -141,6 +143,28 @@ def create_endpoint(self,
             self._controller.create_endpoint.remote(
                 endpoint_name, {backend: 1.0}, route, upper_methods))
 
+        # Block until the route table has been propagated to all HTTP proxies.
+        if route is not None:
+
+            def check_ready(http_response):
```
Comment (simon-mo) [inline python/ray/serve/api.py:149]: This isn't called anywhere.

**Review by simon-mo — APPROVED** [review]
LGTM except in think the check_route isn't actually called, please take a look and fix that before merging.

File: python/ray/serve/api.py (hunk 3 above)
Comment (edoakes) [reply python/ray/serve/api.py:149]: ope, thanks!

**Pull request merged**